File: tgboost/sampling.py
import random
import numpy as np


class RowSampler(object):
    def __init__(self, n, sampling_rate=0.8):
        self.row_mask = np.ones((n,), dtype=bool)
        inds = np.random.choice(n, int(n*(1-sampling_rate)), replace=False)
        self.row_mask[inds] = False

    def shuffle(self):
        np.random.shuffle(self.row_mask)


class ColumnSampler(object):
    def __init__(self, n, sampling_rate=0.8):
        self.col_index = range(n)
        self.n_selected = int(n*sampling_rate)
        self.col_selected = self.col_index[:self.n_selected]

    def shuffle(self):
        random.shuffle(self.col_index)
        self.col_selected = self.col_index[:self.n_selected]

File: src/main/Sampling.java
package main;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class RowSampler{
    public ArrayList<Double> row_mask = new ArrayList<>();

    public RowSampler(int n, double sampling_rate){
        for(int i=0;i<n;i++){
            this.row_mask.add(Math.random()<=sampling_rate? 1.0 : 0.0);
        }
    }

    public void shuffle(){
        Collections.shuffle(this.row_mask);
    }
}


class ColumnSampler{
    private ArrayList<Integer> cols = new ArrayList<>();
    public List<Integer> col_selected;
    private int n_selected;

    public ColumnSampler(int n, double sampling_rate){
        for(int i=0;i<n;i++){
            cols.add(i);
        }
        n_selected = (int) (n * sampling_rate);
        col_selected = cols.subList(0,n_selected);
    }

    public void shuffle(){
        Collections.shuffle(cols);
        col_selected = cols.subList(0,n_selected);
    }
}

public class Sampling {
    public static void main(String[] args) {
        //test case
        RowSampler rs = new RowSampler(1000000, 0.8);
        System.out.println(rs.row_mask.subList(0,20));
        rs.shuffle();
        System.out.println(rs.row_mask.subList(0,20));
        int sum = 0;
        for(double v:rs.row_mask){
            sum += v;
        }
        System.out.println(sum);

        ColumnSampler cs = new ColumnSampler(1000, 0.6);
        System.out.println(cs.col_selected.subList(0,20));
        cs.shuffle();
        System.out.println(cs.col_selected.subList(0,20));
        System.out.println(cs.col_selected.size());
    }
}